services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/service/MongoDbBasedRegistrationService.java

```diff
@@ -196,32 +198,32 @@ public void setConfig(final MongoDbBasedRegistrationConfigProperties config) {
         return processResolveGroupMembers(tenantId, viaGroups, span);
     }
 
-    private JsonObject convertDevice(final String deviceId, final Device payload) {
-
-        if (payload == null) {
-            return null;
-        }
-
-        final JsonObject data = JsonObject.mapFrom(payload);
-
-        return new JsonObject()
-                .put(RegistryManagementConstants.FIELD_PAYLOAD_DEVICE_ID, deviceId)
-                .put("data", data);
+    private Future<DeviceDto> findDevice(final String tenantId, final String deviceId) {
+        return findDeviceDocument(tenantId, deviceId)
+                .compose(result -> Optional.ofNullable(result)
+                        .map(ok -> result.mapTo(DeviceDto.class))
+                        .map(Future::succeededFuture)
+                        .orElseGet(() -> Future.failedFuture(new ClientErrorException(HttpURLConnection.HTTP_NOT_FOUND,
+                                String.format("Device [%s] not found.", deviceId)))));
     }
 
-    private Future<DeviceDto> findDevice(final String tenantId, final String deviceId) {
+    private Future<JsonObject> findDeviceDocument(final String tenantId, final String deviceId) {
         final JsonObject findDeviceQuery = new MongoDbDocumentBuilder()
                 .withTenantId(tenantId)
                 .withDeviceId(deviceId)
                 .document();
         final Promise<JsonObject> readDevicePromise = Promise.promise();
         mongoClient.findOne(config.getCollectionName(), findDeviceQuery, null, readDevicePromise);
-        return readDevicePromise.future()
-                .compose(result -> Optional.ofNullable(result)
-                        .map(ok -> result.mapTo(DeviceDto.class))
-                        .map(Future::succeededFuture)
-                        .orElseGet(() -> Future.failedFuture(new ClientErrorException(HttpURLConnection.HTTP_NOT_FOUND,
-                                String.format("Device [%s] not found.", deviceId)))));
+        return readDevicePromise.future();
```

Comment:
why not fail the promise with a `ClientErrorException(404)` here instead of checking for a `null` result everywhere else?

Reply:
Initially I implemented as in the comment (as below), then I changed it. In this case a `recover` block is needed and there it need to be checked for the _ClientErrorException_ and create a `RegistrationResult` based on the status code. I thought in the other approach, by checking for _null_ we could directly create a `RegistrationResult` with `HTTP_NOT_FOUND` and this extra check for _ClientErrorException_ could be avoided. That's why I chose the other one. We discuss here about the error case and supposedly it's not the often occurring case, I am fine with both the approaches. I will replace it with the below code.


```
return findDeviceDocument(deviceKey.getTenantId(), deviceKey.getDeviceId())
		.map(result -> getRegistrationResult(deviceKey.getDeviceId(),
				result.getJsonObject(MongoDbDeviceRegistryUtils.FIELD_DEVICE)))
		.recover(error -> {
			log.error(error.getMessage(), error);
			TracingHelper.logError(span, error.getMessage(), error);
			if (error instanceof ClientErrorException) {
				return Future.succeededFuture(
						RegistrationResult.from(((ClientErrorException) error).getErrorCode()));
			}
			return Future.succeededFuture(RegistrationResult.from(HttpURLConnection.HTTP_INTERNAL_ERROR));
		});
```

Reply:
No, please keep your code. You seem to have a good reason for doing it the way you do :-)